From a72542f1d8a861da98a62e48b49635cc97246239 Mon Sep 17 00:00:00 2001
From: Riddle Hsu <riddlehsu@google.com>
Date: Mon, 19 May 2025 15:07:16 +0800
Subject: [PATCH 07/25] Use consistent animation when forcibly hiding non
 system overlay

There are several non system overlay window types. When hiding the
windows with animation, their specified exit animations may be used.

This avoids ugly jump cut and inconsistent duration when these
windows are requested to be hidden.

Bug: 408215749
Flag: EXEMPT bugfix
Test: Show multiple windows with different overlay types and
      window animations by non system apps. Launch Settings
      and check the windows are fading out in a short time.
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:dd71ab6ba3950978bd28dc38fecec236cf1153ea)
Merged-In: Ie7bb2663892d608715aa077e2170eae4c03a4e36
Change-Id: Ie7bb2663892d608715aa077e2170eae4c03a4e36
---
 services/core/java/com/android/server/wm/WindowState.java    | 4 ++++
 .../core/java/com/android/server/wm/WindowStateAnimator.java | 5 +++++
 2 files changed, 9 insertions(+)

diff --git a/services/core/java/com/android/server/wm/WindowState.java b/services/core/java/com/android/server/wm/WindowState.java
index 3fa1130d86a3..35d13a54feb9 100644
--- a/services/core/java/com/android/server/wm/WindowState.java
+++ b/services/core/java/com/android/server/wm/WindowState.java
@@ -3159,6 +3159,10 @@ class WindowState extends WindowContainer<WindowState> implements WindowManagerP
         return true;
     }
 
+    boolean isForceHiddenNonSystemOverlayWindow() {
+        return mForceHideNonSystemOverlayWindow;
+    }
+
     void setForceHideNonSystemOverlayWindowIfNeeded(boolean forceHide) {
         final int baseType = getBaseType();
         if (mSession.mCanAddInternalSystemWindow
diff --git a/services/core/java/com/android/server/wm/WindowStateAnimator.java b/services/core/java/com/android/server/wm/WindowStateAnimator.java
index 0154d95d888d..2363a572c3a1 100644
--- a/services/core/java/com/android/server/wm/WindowStateAnimator.java
+++ b/services/core/java/com/android/server/wm/WindowStateAnimator.java
@@ -63,6 +63,7 @@ import android.view.Surface.OutOfResourcesException;
 import android.view.SurfaceControl;
 import android.view.WindowManager;
 import android.view.WindowManager.LayoutParams;
+import android.view.animation.AlphaAnimation;
 import android.view.animation.Animation;
 import android.view.animation.AnimationUtils;
 
@@ -611,6 +612,10 @@ class WindowStateAnimator {
                     a = AnimationUtils.loadAnimation(mContext, anim);
                     Trace.traceEnd(Trace.TRACE_TAG_WINDOW_MANAGER);
                 }
+            } else if (!isEntrance && mWin.isForceHiddenNonSystemOverlayWindow()) {
+                a = new AlphaAnimation(1f, 0f);
+                a.setDuration(mContext.getResources().getInteger(
+                        com.android.internal.R.integer.config_shortAnimTime));
             } else {
                 switch (transit) {
                     case WindowManagerPolicy.TRANSIT_ENTER:
-- 
2.34.1

